README.rst
Splinter plugin for the py.test runner
======================================

.. image:: https://api.travis-ci.org/pytest-dev/pytest-splinter.png
    :target: https://travis-ci.org/pytest-dev/pytest-splinter
.. image:: https://pypip.in/v/pytest-splinter/badge.png
    :target: https://crate.io/packages/pytest-splinter/
.. image:: https://coveralls.io/repos/pytest-dev/pytest-splinter/badge.png?branch=master
    :target: https://coveralls.io/r/pytest-dev/pytest-splinter
.. image:: https://readthedocs.org/projects/pytest-splinter/badge/?version=latest
    :target: https://readthedocs.org/projects/pytest-splinter/?badge=latest
    :alt: Documentation Status


Install pytest-splinter
-----------------------

::

    pip install pytest-splinter


Features
--------

The plugin provides a set of fixtures to use `splinter <http://splinter.readthedocs.org>`_
for browser testing with `pytest <http://pytest.org>`_


Fixtures
--------

* browser
    Get the splinter's Browser. Fixture is underneath session scoped, so browser process is started
    once per test session, but the state of the browser will be clean (current page is ``blank``, cookies clean).
    Note that full cookies clean is only supported for `firefox` webdriver only (for other webdrivers, only cookies
    of the currently visited domain will be cleared). Selenium does not provide direct
    interfaces for cleaning up the cookies of the browser, so for firefox pytest-splinter just removes the
    `cookies.sqlite` file stored in a profile directory. For other drivers, it should be possible to clean all
    cookies as well, this is to be implemented.

* session_browser
    The same as ``browser`` except the lifetime. This fixture is session-scoped so will only be finalized at the
    end of the whole test session. Useful if you want to speedup your test suite paying with reduced test isolation.

* browser_instance_getter
    Function to create an instance of the browser. This fixture is required only if you need to have
    multiple instances of the Browser in a single test at the same time. Example of usage:

.. code-block:: python

    @pytest.fixture
    def admin_browser(request, browser_instance_getter):
        """Admin browser fixture."""
        # browser_instance_getter function receives parent fixture -- our admin_browser
        return browser_instance_getter(request, admin_browser)

    def test_2_browsers(browser, admin_browser):
        """Test using 2 browsers at the same time."""
        browser.visit('http://google.com')
        admin_browser.visit('http://admin.example.com')

* splinter_selenium_implicit_wait
    Implicit wait timeout to be passed to Selenium webdriver.
    Fixture gets the value from the command-line option splinter-implicit-wait (see below)

* splinter_wait_time
    Explicit wait timeout (for waiting for expicit condition via `wait_for_condition`).
    Fixture gets the value from the command-line option splinter-wait-time (see below)

* splinter_selenium_speed
    Speed for Selenium, if not 0 then it will sleep between each selenium command.
    Useful for debugging/demonstration.
    Fixture gets the value from the command-line option splinter-speed (see below)

* splinter_selenium_socket_timeout
    Socket timeout for communication between the webdriver and the browser.
    Fixture gets the value from the command-line option splinter-socket-timeout (see below)

* splinter_webdriver
    Splinter's webdriver name to use. Fixture gets the value from the command-line option
    splinter-webdriver (see below). To make pytest-splinter always use certain webdriver, override a fixture
    in your `conftest.py` file:

.. code-block:: python

    import pytest

    @pytest.fixture(scope='session')
    def splinter_webdriver():
        """Override splinter webdriver name."""
        return 'phantomjs'

* splinter_remote_url
    Splinter's webdriver remote url to use (optional). Fixture gets the value from the command-line option
    splinter-remote-url (see below). Will be used only if selected webdriver name is 'remote'.

* splinter_session_scoped_browser
    pytest-splinter should use single browser instance per test session.
    Fixture gets the value from the command-line option splinter-session-scoped-browser (see below)

* splinter_file_download_dir
    Directory, to which browser will automatically download the files it
    will experience during browsing. For example when you click on some download link.
    By default it's a temporary directory. Automatic downloading of files is only supported for firefox driver
    at the moment.

* splinter_download_file_types
    Comma-separated list of content types to automatically download.
    By default it's the all known system mime types (via mimetypes standard library).

* splinter_browser_load_condition
    Browser load condition, python function which should return True.
    If function returns False, it will be run several times, until timeout below reached.

* splinter_browser_load_timeout
    Browser load condition timeout in seconds, after this timeout the exception
    WaitUntilTimeout will be raised.

* splinter_wait_time
    Browser explicit wait timeout in seconds, after this timeout the exception
    WaitUntilTimeout will be raised.

* splinter_firefox_profile_preferences
    Firefox profile preferences, a dictionary which is passed to selenium
    webdriver's profile_preferences

* splinter_firefox_profile_directory
    Firefox profile directory to use as template for firefox profile created by selenium.
    By default, it's an empty directly inside pytest_splinter/profiles/firefox

* splinter_driver_kwargs
    Webdriver keyword arguments, a dictionary which is passed to selenium
    webdriver's constructor (after applying firefox preferences)

* splinter_window_size
    Size of the browser window on browser initialization. Tuple in form (<width>, <height>). Default is (1366, 768)

* splinter_screenshot_dir
    pytest-splinter browser screenshot directory.
    This fixture gets the value from the command-line option
    `splinter-screenshot-dir` (see below).

* splinter_make_screenshot_on_failure
    Should pytest-splinter take browser screenshots on test failure?
    This fixture gets the value from the command-line option
    `splinter-make-screenshot-on-failure` (see below).

* splinter_screenshot_encoding
    Encoding of the `html` `screenshot` on test failure. UTF-8 by default.

* splinter_screenshot_getter_html
    Function to get browser html screenshot. By default, it saves `browser.html` with given path and
    `splinter_screenshot_encoding` encoding.

* splinter_screenshot_getter_png
    Function to get browser image (png) screenshot. By default, it calls `browser.save_sceenshot`
    with given path.

* splinter_driver_executable
    Filesystem path of the webdriver executable.
    This fixture gets the value from the command-line option
    `splinter-webdriver-executable` (see below).

* splinter_browser_class
    Class to use for browser instance.
    Defaults to `pytest_splinter.plugin.Browser`.


Command-line options
--------------------

* `--splinter-implicit-wait`
    Selenium webdriver implicit wait. Seconds (default: 5).

* `--splinter-speed`
    selenium webdriver speed (from command to command). Seconds (default: 0).

* `--splinter-socket-timeout`
    Selenium webdriver socket timeout for for communication between the webdriver and the browser.
    Seconds (default: 120).

* `--splinter-webdriver`
    Webdriver name to use. (default: firefox). Options:

    *  firefox
    *  remote
    *  chrome
    *  phantomjs

    For more details refer to the documentation for splinter and selenium.

* `--splinter-remote-url`
    Webdriver remote url to use. (default: None). Will be used only if selected webdriver name is 'remote'.

    For more details refer to the documentation for splinter and selenium.

* `--splinter-session-scoped-browser`
    pytest-splinter should use a single browser instance per test session.
    Choices are 'true' or 'false' (default: 'true').

* `--splinter-make-screenshot-on-failure`
    pytest-splinter should take browser screenshots on test failure.
    Choices are 'true' or 'false' (default: 'true').

* `--splinter-screenshot-dir`
    pytest-splinter browser screenshot directory. Defaults to the current
    directory.

* `--splinter-webdriver-executable`
    Filesystem path of the webdriver executable. Used by phantomjs and chrome drivers.
    Defaults to the None in which case the shell PATH variable setting determines the location of the executable.


Browser fixture
---------------

As mentioned above, browser is a fixture made by creating splinter's Browser object, but with some overrides.

*  visit
    Added possibility to wait for condition on each browser visit by having a fixture.

*  wait_for_condition
    Method copying selenium's wait_for_condition, with difference that condition is in python,
    so there you can do whatever you want, and not only execute javascript via browser.evaluate_script.


Several browsers for your test
------------------------------

You can have several browsers in one test.

.. code-block:: python

    import pytest

    @pytest.fixture
    def admin_browser(browser_instance_getter):
        return browser_instance_getter(admin_browser)

    def test_with_several_browsers(browser, admin_browser):
        browser.visit('http://example.com')
        admin_browser.visit('about:blank')
        assert browser.url == 'http://example.com'


Automatic screenshots on test failure
-------------------------------------

When your functional test fails, it's important to know the reason.
This becomes hard when tests are being run on the continuos integration server,
where you cannot debug (using --pdb).
To simplify things, a special behaviour of the browser fixture is available,
which takes a screenshot on test failure and puts it in a folder with the a
naming convention compatible to the
`jenkins plugin <https://wiki.jenkins-ci.org/display/JENKINS/JUnit+Attachments+Plugin>`_.
The html content of the browser page is also stored, this can be useful for debugging the html source.

Creating screenshots is fully compatible with `pytest-xdist plugin
<https://pypi.python.org/pypi/pytest-xdist>`_ and will transfer the screenshots
from the slave nodes through the communication channel automatically.

If a test (using the browser fixture) fails, you should get a screenshot files
in the following path:

::

    <splinter-screenshot-dir>/my.dotted.name.test.package/test_name-browser.png
    <splinter-screenshot-dir>/my.dotted.name.test.package/test_name-browser.html

The `splinter-screenshot-dir` for storing the screenshot is generated by a
fixture and can be provided through a command line argument, as described above
at the configuration options section.

Taking screenshots on test failure is enabled by default. It can be controlled
through the `splinter_make_screenshot_on_failure` fixture, where return `False`
skips it. You can also disable it via a command line argument:

::

    py.test tests/functional --splinter-make-screenshot-on-failure=false

In case taking a screenshot fails, a pytest warning will be issued, which
can be viewed using the `-rw` argument for `py.test`.


Python3 support
---------------

Python3 is supported, check if you have recent version of splinter as it was added recently.


Example
-------

test_your_test.py:

.. code-block:: python

    def test_some_browser_stuff(browser):
        """Test using real browser."""
        url = "http://www.google.com"
        browser.visit(url)
        browser.fill('q', 'splinter - python acceptance testing for web applications')
        # Find and click the 'search' button
        button = browser.find_by_name('btnK')
        # Interact with elements
        button.click()
        assert browser.is_text_present('splinter.cobrateam.info'), 'splinter.cobrateam.info wasn't found... We need to'
        ' improve our SEO techniques'


Contact
-------

If you have questions, bug reports, suggestions, etc. please create an issue on
the `GitHub project page <http://github.com/paylogic/pytest-splinter>`_.


License
-------


CHANGES.rst
Changelog
=========

1.6.7
-----

- clean cookies file between tests for firefox (bubenkoff)

1.6.6
-----

- screenshot encoding made flexible (bubenkoff)

1.6.2
-----

- pass timeout to restored connection (bubenkoff)

1.6.0
-----

- added html screenshot (bubenkoff, blueyed)

1.5.3
-----

- remote webdriver fixes (bubenkoff)

1.5.2
-----

- respect splinter_make_screenshot_on_failure (bubenkoff)

1.5.1
-----

- use native selenium socket timeout feature (pelme)

1.5.0
-----

- pytest tmpdir_factory support (bubenkoff)
- depend on splinter 0.7.3, remove the previous status_code monkey patch (pelme)
- add option `--splinter-wait-time` to specify splinter explicit wait timeout (pelme)

1.4.6
-----

- ensure base tempdir exists (bubenkoff)


1.4.0
-----

- introduce splinter_browser_class fixture (bubenkoff, ecesena)


1.3.8
-----

- correctly handle zope.testbrowser splinter driver (bubenkoff)


1.3.7
-----

- pass `splinter_selenium_implicit_wait` as `wait_time` to splinter Browser (lrowe)


1.3.6
-----

- properly respect webdriver executable command line option (bubenkoff, bh)


1.3.5
-----

- add option --splinter-webdriver-executable for phantomjs and chrome (sureshvv)


1.3.4
-----

- make ``browser_instance_getter`` session scoped, add ``session_browser`` fixture (bubenkoff, sureshvv)


1.3.3
-----

- make ``mouse_over`` comparible with more use-cases (bubenkoff)


1.3.1
-----

- properly handle driver switch during the test run (bubenkoff)
- respect splinter_session_scoped_browser fixture (bubenkoff)


1.2.10
------

- handle exceptions during screenshot saving (blueyed, bubenkoff)
- documentation improvements (blueyed)


1.2.9
-----

- status_code is back in a lazy way (bubenkoff)


1.2.7
-----

- Fix automatic download of pdf content type (bubenkoff)


1.2.4
-----

- fix failing the test run if pytest-xdist is not installed, as it's completely optional dependency (bubenkoff, slafs)


1.2.3
-----

- improve exception handing when preparing the browser instance (bubenkoff)
- require pytest (bubenkoff)


1.2.0
-----

- automatic screenshot capture on test failure (bubenkoff)
- improvements to the browser preparation procedure (bubenkoff)
- boolean config options made more clear (bubenkoff)


1.1.1
-----

- restore browser parameters on each test run instead of once for browser start (bubenkoff)


1.1.0
-----

- added possibility to have multiple browser instances for single test (amakhnach, bubenkoff)


1.0.4
-----

- Fixed browser fixture to support splinter_browser_load_condition and splinter_browser_load_timeout by default. (markon)


1.0.3
-----

- unicode fixes to setup.py (bubenkoff, valberg)


1.0.2
-----

- wait_for_condition now receives pytest_bdd.plugin.Browser object, not selenium webdriver one (bubenkoff)


1.0.1
-----

- Refactoring and cleanup (bubenkoff)


1.0.0
-----

- Initial public release


pytest_splinter/plugin.py
"""Splinter plugin for pytest.

Provides easy interface for the browser from your tests providing the `browser` fixture
which is an object of splinter Browser class.
"""
import codecs
import functools  # pragma: no cover
try:
    from httplib import HTTPException
except ImportError:
    from http.client import HTTPException
import mimetypes  # pragma: no cover
import os.path
import re
from os import unlink

import pytest  # pragma: no cover
import splinter  # pragma: no cover
from _pytest import junitxml
from _pytest.tmpdir import tmpdir

from selenium.webdriver.support import wait
from selenium.common.exceptions import WebDriverException

from .webdriver_patches import patch_webdriver  # pragma: no cover
from .splinter_patches import patch_webdriverelement  # pragma: no cover
from .decorators import with_fixtures

import logging
LOGGER = logging.getLogger(__name__)


NAME_RE = re.compile(r'[\W]')


def _visit(self, url):
    """Override splinter's visit to avoid unnecessary checks and add wait_until instead."""
    self.driver.get(url)
    self.wait_for_condition(self.visit_condition, timeout=self.visit_condition_timeout)


def _wait_for_condition(self, condition=None, timeout=None, poll_frequency=0.5, ignored_exceptions=None):
    """Wait for given javascript condition."""
    condition = functools.partial(condition or self.visit_condition, self)

    timeout = timeout or self.wait_time

    return wait.WebDriverWait(
        self.driver, timeout, poll_frequency=poll_frequency, ignored_exceptions=ignored_exceptions
    ).until(
        lambda browser: condition()
    )


def Browser(*args, **kwargs):
    """Emulate splinter's Browser."""
    visit_condition = kwargs.pop('visit_condition')
    visit_condition_timeout = kwargs.pop('visit_condition_timeout')
    browser = splinter.Browser(*args, **kwargs)
    browser.wait_for_condition = functools.partial(_wait_for_condition, browser)
    if hasattr(browser, 'driver'):
        browser.visit_condition = visit_condition
        browser.visit_condition_timeout = visit_condition_timeout
        browser.visit = functools.partial(_visit, browser)
    browser.__splinter_browser__ = True
    return browser


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_close_browser():
    """Close browser fixture."""
    return True


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_webdriver(request):
    """Webdriver fixture."""
    return request.config.option.splinter_webdriver


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_remote_url(request):
    """Remote webdriver url.

    :return: URL of remote webdriver.
    """
    return request.config.option.splinter_remote_url


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_selenium_socket_timeout(request):
    """Internal Selenium socket timeout (communication between webdriver and the browser).

    :return: Seconds.
    """
    return request.config.option.splinter_webdriver_socket_timeout


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_selenium_implicit_wait(request):
    """Selenium implicit wait timeout.

    :return: Seconds.
    """
    return request.config.option.splinter_webdriver_implicit_wait


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_wait_time(request):
    """Splinter explicit wait timeout.

    :return: Seconds.
    """
    return request.config.option.splinter_wait_time


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_selenium_speed(request):
    """Selenium speed.

    :return: Seconds.
    """
    return request.config.option.splinter_webdriver_speed


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_browser_load_condition():
    """The condition that has to be `True` to assume that the page is fully loaded.

    One example is to wait for jQuery, then the condition could be::

        @pytest.fixture
        def splinter_browser_load_condition():

            def condition(browser):
                return browser.evaluate_script('typeof $ === "undefined" || !$.active')

            return condition
    """
    return lambda browser: True


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_browser_load_timeout():
    """The timeout in seconds in which the page is expected to be fully loaded."""
    return 10


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_file_download_dir(session_tmpdir):
    """Browser file download directory."""
    return session_tmpdir.ensure('splinter', 'download', dir=True).strpath


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_download_file_types():
    """Browser file types to download. Comma-separated."""
    return ','.join(mimetypes.types_map.values())


@pytest.fixture(scope='session')
def splinter_firefox_profile_preferences():
    """Firefox profile preferences."""
    return {
        'browser.cache.memory.enable': False,
        'browser.sessionhistory.max_total_viewers': 0,
        'network.http.pipelining': True,
        'network.http.pipelining.maxrequests': 8
    }


@pytest.fixture(scope='session')
def splinter_firefox_profile_directory(session_tmpdir):
    """Firefox profile directory."""
    path = session_tmpdir.join('profiles', 'firefox').ensure(dir=True)
    return path.strpath


def clean_cookies(browser, webdriver_name):
    """Clean browser cookies."""
    browser.cookies.delete()
    if webdriver_name == 'firefox':
        path = os.path.join(browser.driver.profile.profile_dir, 'cookies.sqlite')
        if os.path.exists(path):
            unlink(path)


@pytest.fixture(scope='session')
def splinter_driver_kwargs():
    """Webdriver kwargs."""
    return {}


@pytest.fixture(scope='session')
def splinter_window_size():
    """Browser window size. (width, height)."""
    return (1366, 768)


@pytest.fixture(scope='session')
def splinter_session_scoped_browser(request):
    """Flag to keep single browser per test session."""
    return request.config.option.splinter_session_scoped_browser == 'true'


@pytest.fixture(scope='session')
def splinter_make_screenshot_on_failure(request):
    """Flag to make browser screenshot on test failure."""
    return request.config.option.splinter_make_screenshot_on_failure == 'true'


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_screenshot_dir(request):
    """Browser screenshot directory."""
    return os.path.abspath(request.config.option.splinter_screenshot_dir)


@pytest.fixture(scope='session')  # pragma: no cover
def splinter_screenshot_encoding(request):
    """Browser screenshot html encoding."""
    return 'utf-8'


@pytest.fixture(scope='session')
def splinter_webdriver_executable(request):
    """Webdriver executable directory."""
    executable = request.config.option.splinter_webdriver_executable
    return os.path.abspath(executable) if executable else None


@pytest.fixture(scope='session')
def browser_pool(request, splinter_close_browser):
    """Browser 'pool' to emulate session scope but with possibility to recreate browser."""
    pool = {}

    def fin():
        for browser in pool.values():
            try:
                browser.quit()
            except Exception:  # NOQA
                pass

    if splinter_close_browser:
        request.addfinalizer(fin)

    return pool


@pytest.fixture(scope='session')
def browser_patches():
    """Browser monkey patches."""
    patch_webdriver()
    patch_webdriverelement()


@pytest.fixture(scope='session')
def session_tmpdir(request):
    """pytest tmpdir which is session-scoped."""
    return with_fixtures(tmpdir)(request)


@pytest.fixture(scope='session')
def splinter_browser_class(request):
    """Browser class to use for browser instance creation."""
    return Browser


def get_args(driver=None,
             download_dir=None,
             download_ftypes=None,
             firefox_pref=None,
             firefox_prof_dir=None,
             remote_url=None,
             executable=None,
             driver_kwargs=None):
    """Construct arguments to be passed to webdriver on initialization."""
    kwargs = {}

    if driver == 'firefox':
        kwargs['profile_preferences'] = dict({
            'browser.download.folderList': 2,
            'browser.download.manager.showWhenStarting': False,
            'browser.download.dir': download_dir,
            'browser.helperApps.neverAsk.saveToDisk': download_ftypes,
            'browser.helperApps.alwaysAsk.force': False,
            'pdfjs.disabled': True,  # disable internal ff pdf viewer to allow auto pdf download
        }, **firefox_pref)
        kwargs['profile'] = firefox_prof_dir
    elif driver == 'remote':
        if remote_url:
            kwargs['url'] = remote_url
        kwargs['keep_alive'] = True
    elif driver in ('phantomjs', 'chrome'):
        if executable:
            kwargs['executable_path'] = executable
    if driver_kwargs:
        kwargs.update(driver_kwargs)

    return kwargs


@pytest.fixture(scope='session')
def splinter_screenshot_getter_png():
    """Screenshot getter function: png."""
    def getter(browser, path):
        browser.driver.save_screenshot(path)
    return getter


@pytest.fixture(scope='session')
def splinter_screenshot_getter_html(splinter_screenshot_encoding):
    """Screenshot getter function: html."""
    def getter(browser, path):
        with codecs.open(path, 'w', encoding=splinter_screenshot_encoding) as fd:
            fd.write(browser.html)
    return getter


@pytest.fixture(scope='session')
def browser_instance_getter(
        browser_patches,
        splinter_browser_load_condition,
        splinter_browser_load_timeout,
        splinter_download_file_types,
        splinter_driver_kwargs,
        splinter_file_download_dir,
        splinter_firefox_profile_preferences,
        splinter_firefox_profile_directory,
        splinter_make_screenshot_on_failure,
        splinter_remote_url,
        splinter_screenshot_dir,
        splinter_selenium_implicit_wait,
        splinter_wait_time,
        splinter_selenium_socket_timeout,
        splinter_selenium_speed,
        splinter_webdriver_executable,
        splinter_window_size,
        splinter_browser_class,
        session_tmpdir,
        browser_pool,
):
    """Splinter browser instance getter. To be used for getting of plugin.Browser's instances.

    :return: function(parent). Each time this function will return new instance of plugin.Browser class.
    """
    def get_browser(splinter_webdriver, retry_count=3):
        kwargs = get_args(driver=splinter_webdriver,
                          download_dir=splinter_file_download_dir,
                          download_ftypes=splinter_download_file_types,
                          firefox_pref=splinter_firefox_profile_preferences,
                          firefox_prof_dir=splinter_firefox_profile_directory,
                          remote_url=splinter_remote_url,
                          executable=splinter_webdriver_executable,
                          driver_kwargs=splinter_driver_kwargs)
        try:
            return splinter_browser_class(
                splinter_webdriver, visit_condition=splinter_browser_load_condition,
                visit_condition_timeout=splinter_browser_load_timeout,
                wait_time=splinter_wait_time, **kwargs
            )
        except Exception:  # NOQA
            if retry_count > 1:
                return get_browser(splinter_webdriver, retry_count - 1)
            else:
                raise

    def prepare_browser(request, parent):
        splinter_webdriver = request.getfuncargvalue('splinter_webdriver')
        splinter_session_scoped_browser = request.getfuncargvalue('splinter_session_scoped_browser')
        browser_key = id(parent)
        browser = browser_pool.get(browser_key)
        if not splinter_session_scoped_browser:
            browser = get_browser(splinter_webdriver)
            if splinter_close_browser:
                request.addfinalizer(browser.quit)
        elif not browser:
            browser = browser_pool[browser_key] = get_browser(splinter_webdriver)
        try:
            if splinter_webdriver not in browser.driver_name.lower():
                raise IOError('webdriver does not match')
            if hasattr(browser, 'driver'):
                browser.driver.implicitly_wait(splinter_selenium_implicit_wait)
                browser.driver.set_speed(splinter_selenium_speed)
                browser.driver.command_executor.set_timeout(splinter_selenium_socket_timeout)
                _conn = getattr(browser.driver.command_executor, '_conn', None)
                if _conn:
                    _conn.timeout = splinter_selenium_socket_timeout
                if splinter_window_size:
                    browser.driver.set_window_size(*splinter_window_size)
            clean_cookies(browser, splinter_webdriver)
            if hasattr(browser, 'driver'):
                browser.visit_condition = splinter_browser_load_condition
                browser.visit_condition_timeout = splinter_browser_load_timeout
                browser.visit('about:blank')
        except (IOError, HTTPException, WebDriverException):
            # we lost browser, try to restore the justice
            try:
                browser.quit()
            except Exception:  # NOQA
                pass
            browser = browser_pool[browser_key] = get_browser(splinter_webdriver)
            prepare_browser(request, parent)

        return browser

    return prepare_browser


@pytest.yield_fixture(autouse=True)
def browser_screenshot(
        request, splinter_screenshot_dir, session_tmpdir, splinter_make_screenshot_on_failure,
        splinter_screenshot_encoding, splinter_screenshot_getter_png, splinter_screenshot_getter_html):
    """Make browser screenshot on test failure."""
    yield
    for name, value in request._funcargs.items():
        if hasattr(value, '__splinter_browser__'):
            browser = value
            if splinter_make_screenshot_on_failure and request.node.splinter_failure:
                slaveoutput = getattr(request.config, 'slaveoutput', None)
                names = junitxml.mangle_testnames(request.node.nodeid.split("::"))
                classname = '.'.join(names[:-1])
                screenshot_dir = os.path.join(splinter_screenshot_dir, classname)
                screenshot_file_name_format = '{0}-{1}.{{format}}'.format(
                    names[-1][:128 - len(name) - 5], name)
                screenshot_file_name = screenshot_file_name_format.format(format='png')
                screenshot_html_file_name = screenshot_file_name_format.format(format='html')
                if not slaveoutput:
                    if not os.path.exists(screenshot_dir):
                        os.makedirs(screenshot_dir)
                else:
                    screenshot_dir = session_tmpdir.ensure('screenshots', dir=True).strpath
                screenshot_png_path = os.path.join(screenshot_dir, screenshot_file_name)
                screenshot_html_path = os.path.join(screenshot_dir, screenshot_html_file_name)
                LOGGER.info('Saving screenshot to %s', screenshot_dir)
                try:
                    splinter_screenshot_getter_html(browser, screenshot_html_path)
                    splinter_screenshot_getter_png(browser, screenshot_png_path)
                    if slaveoutput is not None:
                        with codecs.open(screenshot_html_path, encoding=splinter_screenshot_encoding) as html_fd:
                            with open(screenshot_png_path) as fd:
                                slaveoutput.setdefault('screenshots', []).append({
                                    'class_name': classname,
                                    'files': [
                                        {
                                            'file_name': screenshot_file_name,
                                            'content': fd.read(),
                                        },
                                        {
                                            'file_name': screenshot_html_file_name,
                                            'content': html_fd.read(),
                                            'encoding': splinter_screenshot_encoding
                                        }]
                                })
                except Exception as e:  # NOQA
                    request.config.warn('SPL504', "Could not save screenshot: {0}".format(e))


@pytest.mark.tryfirst
def pytest_runtest_makereport(item, call, __multicall__):
    """Assign the report to the item for futher usage."""
    rep = __multicall__.execute()
    if rep.outcome != 'passed':
        item.splinter_failure = rep
    else:
        item.splinter_failure = None
    return rep


@pytest.fixture
def browser(request, splinter_session_scoped_browser, splinter_webdriver, browser_instance_getter):
    """Browser fixture."""
    return browser_instance_getter(request, browser)


@pytest.fixture(scope='session')
def session_browser(request, browser_instance_getter):
    """Session scoped browser fixture."""
    return browser_instance_getter(request, session_browser)


class SplinterXdistPlugin(object):

    """Plugin class to defer pytest-xdist hook handler."""

    def pytest_testnodedown(self, node, error):
        """Copy screenshots back from remote nodes to have them on the master."""
        config_screenshot_dir = splinter_screenshot_dir(node)
        for screenshot in getattr(node, 'slaveoutput', {}).get('screenshots', []):
            screenshot_dir = os.path.join(config_screenshot_dir, screenshot['class_name'])
            if not os.path.exists(screenshot_dir):
                os.makedirs(screenshot_dir)
            for fil in screenshot['files']:
                encoding = fil.get('encoding')
                with codecs.open(os.path.join(screenshot_dir, fil['file_name']), 'w',
                                 **dict(encoding=encoding) if encoding else {}) as fd:
                    fd.write(fil['content'])


def pytest_configure(config):
    """Register pytest-splinter's deferred plugin."""
    if config.pluginmanager.getplugin('xdist'):
        config.pluginmanager.register(SplinterXdistPlugin())


def pytest_addoption(parser):  # pragma: no cover
    """Pytest hook to add custom command line option(s)."""
    group = parser.getgroup("splinter", "splinter integration for browser testing")
    group.addoption(
        "--splinter-webdriver",
        help="pytest-splinter webdriver", type="choice", choices=list(splinter.browser._DRIVERS.keys()),
        dest='splinter_webdriver', metavar="DRIVER", default='firefox')
    group.addoption(
        "--splinter-remote-url",
        help="pytest-splinter remote webdriver url ", metavar="URL", dest='splinter_remote_url', default=None)
    group.addoption(
        "--splinter-wait-time",
        help="splinter explicit wait, seconds", type="int",
        dest='splinter_wait_time', metavar="SECONDS", default=5)
    group.addoption(
        "--splinter-implicit-wait",
        help="pytest-splinter selenium implicit wait, seconds", type="int",
        dest='splinter_webdriver_implicit_wait', metavar="SECONDS", default=5)
    group.addoption(
        "--splinter-speed",
        help="pytest-splinter selenium speed, seconds", type="int",
        dest='splinter_webdriver_speed', metavar="SECONDS", default=0)
    group.addoption(
        "--splinter-socket-timeout",
        help="pytest-splinter socket timeout, seconds", type="int",
        dest='splinter_webdriver_socket_timeout', metavar="SECONDS", default=120)
    group.addoption(
        "--splinter-session-scoped-browser",
        help="pytest-splinter should use a single browser instance per test session. Defaults to true.", action="store",
        dest='splinter_session_scoped_browser', metavar="false|true", type="choice", choices=['false', 'true'],
        default='true')
    group.addoption(
        "--splinter-make-screenshot-on-failure",
        help="pytest-splinter should take browser screenshots on test failure. Defaults to true.", action="store",
        dest='splinter_make_screenshot_on_failure', metavar="false|true", type="choice", choices=['false', 'true'],
        default='true')
    group.addoption(
        "--splinter-screenshot-dir",
        help="pytest-splinter browser screenshot directory. Defaults to the current directory.", action="store",
        dest='splinter_screenshot_dir', metavar="DIR", default='.')
    group.addoption(
        "--splinter-webdriver-executable",
        help="pytest-splinter webdrive executable path. Defaults to unspecified in which case it is taken from PATH",
        action="store",
        dest='splinter_webdriver_executable', metavar="DIR", default='')
